<a href="https://colab.research.google.com/github/Feranmi149/OIBSIP/blob/main/Task_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd
df = pd.read_csv('The Titanic dataset.csv')
df.head()

,1,2,3,4,5,6,7,8,9,10
0,sn,pclass,survived,NaN,gender,age,family,fare,embarked,date
1,1,3,0,Mr. Anthony,male,42,0,7.55,NaN,01-Jan-90
2,1,3,0,Mr. Anthony,male,42,0,7.55,NaN,01-Jan-90
3,2,3,0,Master. Eugene Joseph,male,?,2,20.25,S,02-Jan-90
4,3,2,0,"Abbott, Mr. Rossmore Edward",NaN,NaN,2,**,S,03-Jan-90


In [4]:
print("Shape:", df.shape)
print("\nNulls per column:\n", df.isnull().sum())
print("\nDuplicate rows:", df.duplicated().sum())
print("\nData types:\n", df.dtypes)

Shape: (1302, 10)

Nulls per column:
 1       0
2       0
3       0
4       1
5       1
6     257
7       2
8       2
9       6
10      0
dtype: int64

Duplicate rows: 1

Data types:
 1     object
2     object
3     object
4     object
5     object
6     object
7     object
8     object
9     object
10    object
dtype: object


In [5]:
df.columns

Index(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10'], dtype='object')

In [6]:
df.head()

,1,2,3,4,5,6,7,8,9,10
0,sn,pclass,survived,NaN,gender,age,family,fare,embarked,date
1,1,3,0,Mr. Anthony,male,42,0,7.55,NaN,01-Jan-90
2,1,3,0,Mr. Anthony,male,42,0,7.55,NaN,01-Jan-90
3,2,3,0,Master. Eugene Joseph,male,?,2,20.25,S,02-Jan-90
4,3,2,0,"Abbott, Mr. Rossmore Edward",NaN,NaN,2,**,S,03-Jan-90


In [7]:
df = pd.read_csv('The Titanic dataset.csv', header=1)
df.rename(columns={'Unnamed: 3': 'name'}, inplace=True)
df.head()

,sn,pclass,survived,name,gender,age,family,fare,embarked,date
0,1,3,0,Mr. Anthony,male,42,0.0,7.55,NaN,01-Jan-90
1,1,3,0,Mr. Anthony,male,42,0.0,7.55,NaN,01-Jan-90
2,2,3,0,Master. Eugene Joseph,male,?,2.0,20.25,S,02-Jan-90
3,3,2,0,"Abbott, Mr. Rossmore Edward",NaN,NaN,2.0,**,S,03-Jan-90
4,4,3,1,"Abbott, Mr. Rossmore Edward",female,35,2.0,20.25,S,04-Jan-90


In [8]:
print("Shape:", df.shape)
print("\nNulls per column:\n", df.isnull().sum())
print("\nDuplicate rows:", df.duplicated().sum())
print("\nData types:\n", df.dtypes)

Shape: (1301, 10)

Nulls per column:
 sn            0
pclass        0
survived      0
name          0
gender        1
age         257
family        2
fare          2
embarked      6
date          0
dtype: int64

Duplicate rows: 1

Data types:
 sn            int64
pclass        int64
survived      int64
name         object
gender       object
age          object
family      float64
fare         object
embarked     object
date         object
dtype: object


Data Quality Report (Before Cleaning): 1,301 rows, 10 columns. Missing values found in gender (1), age (257), family (2), fare (2), and embarked (6). One duplicate row identified. Age and fare are incorrectly typed as object/text due to placeholder values like "?" and "**" representing missing data.

In [9]:
import numpy as np
df['age'] = df['age'].replace('?', np.nan)
df['fare'] = df['fare'].replace('**', np.nan)

In [10]:
df['age'] = pd.to_numeric(df['age'])
df['fare'] = pd.to_numeric(df['fare'])

In [11]:
df.isnull().sum()

,0
sn,0
pclass,0
survived,0
name,0
gender,1
age,258
family,2
fare,3
embarked,6
date,0


In [12]:
df['gender'] = df['gender'].fillna(df['gender'].mode()[0])

Justification: Gender is categorical with only 1 missing value; filling with the most common value (mode) is reasonable and has negligible impact on the dataset.

In [13]:
df['age'] = df['age'].fillna(df['age'].median())

Justification: Age has a large number of missing values (258). Median is used instead of mean because age can be skewed by outliers (e.g., infants or elderly passengers), and median is more robust to that skew.

In [14]:
df['family'] = df['family'].fillna(df['family'].mode()[0])

Justification: Family size is a count value; only 2 records are missing, so filling with the most common family size (mode) is a safe, low-impact choice.

In [15]:
df['fare'] = df['fare'].fillna(df['fare'].median())

Justification: Fare is monetary and can have outliers (luxury cabins vs. steerage); median avoids skew from extreme high fares better than mean.

In [16]:
df['embarked'] = df['embarked'].fillna(df['embarked'].mode()[0])

Embarkation port is categorical with only 6 missing values; mode imputation is standard practice here since it's low-impact and preserves the most common category.

In [17]:
df.isnull().sum()

,0
sn,0
pclass,0
survived,0
name,0
gender,0
age,0
family,0
fare,0
embarked,0
date,0


In [18]:
print("Duplicates before:", df.duplicated().sum())
df = df.drop_duplicates()
print("Duplicates after:", df.duplicated().sum())
print("New shape:", df.shape)

Duplicates before: 1
Duplicates after: 0
New shape: (1300, 10)


Duplicate Removal: 1 duplicate row was identified and removed (the repeated "Mr. Anthony" record). Row count reduced from 1301 to 1300.

In [19]:
df['gender'].unique()

array(['male', 'female'], dtype=object)

In [20]:
df['embarked'].unique()

array(['S', 'C', 'Q'], dtype=object)

Gender (male/female) and embarked (S, C, Q) columns were already consistently formatted so no changes needed.

In [21]:
df['date'] = pd.to_datetime(df['date'], format='%d-%b-%y')
df.dtypes

,0
sn,int64
pclass,int64
survived,int64
name,object
gender,object
age,float64
family,float64
fare,float64
embarked,object
date,datetime64[ns]


In [22]:
df['sn'] = df['sn'].astype(str)
df.dtypes

,0
sn,object
pclass,int64
survived,int64
name,object
gender,object
age,float64
family,float64
fare,float64
embarked,object
date,datetime64[ns]


Data Type Correction: All columns now have appropriate types, sn converted to string (as an identifier, not a quantity), age/family/fare as float, date as proper datetime, and categorical columns (gender, embarked) remain as text.

In [23]:
Q1 = df['age'].quantile(0.25)
Q3 = df['age'].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR
outliers_age = df[(df['age'] < lower) | (df['age'] > upper)]
print("Age outliers:", outliers_age.shape[0])
print("Valid range:", lower, "to", upper)

Age outliers: 99
Valid range: 2.5 to 54.5


In [24]:
Q1 = df['fare'].quantile(0.25)
Q3 = df['fare'].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR
outliers_fare = df[(df['fare'] < lower) | (df['fare'] > upper)]
print("Fare outliers:", outliers_fare.shape[0])
print("Valid range:", lower, "to", upper)

Fare outliers: 171
Valid range: -27.172999999999995 to 66.34379999999999


Outlier Detection (IQR method): Using the IQR method, 99 age outliers were identified (valid range: 2.5–54.5) and 171 fare outliers were identified (valid range: -27.17 to 66.34). Reviewing the data, age outliers represent genuinely older passengers rather than data entry errors, ship passengers naturally span a wide age range. Fare outliers largely reflect higher-class ticket prices (first-class fares were legitimately much higher than steerage), which is expected given how fare correlates with passenger class rather than being erroneous. Both sets of outliers were retained rather than capped or removed, since removing them would distort the natural variation in passenger age and ticket pricing that the dataset is meant to reflect.

In [26]:
summary = pd.DataFrame({
    'Metric': ['Row Count', 'Null Count', 'Duplicate Count'],
    'Before': [1301, 270, 1],
    'After': [df.shape[0], df.isnull().sum().sum(), df.duplicated().sum()]
})
summary

,Metric,Before,After
0,Row Count,1301,1300
1,Null Count,270,0
2,Duplicate Count,1,0


Before vs. After Summary: Cleaning reduced null values from 270 to 0, removed the single duplicate row, and corrected data types across all columns. The dataset is now fully clean and analysis-ready.

In [27]:
df.to_csv('cleaned_titanic_dataset.csv', index=False)